### PR TITLE
feature: adds redirect map support for VM.require

### PR DESCRIPTION
**Pull request by elliotBraem**
Allows local module development when using VM.require.

Moves computeSrcOrCode to a utility class so it can be shared with VM.require, then checks if code exists in redirect map before doing Social.get

Extension of [Ability to recursively redirect widget's src (#134)](https://github.com/NearSocial/VM/commit/86f85fef97a4ded1a5c8859b0de718806cfc4e44) for VM.require

### CODE REVIEW

File: src/lib/vm/vm.js [hunk 1]
```diff
@@ -2069,16 +2070,24 @@ export default class VM {
...
+        undefined
+      );
+      if (!code) {
+        return code;
+      }
+    } else if (srcOrCode?.code) {
+      code = srcOrCode.code;
     }
+    console.log(code);
     const vm = this.getVmInstance(code, src);
```
Comment (evgenykuzyakov) [inline src/lib/vm/vm.js:2091]: This should be updated based on `srcOrCode?.src`
```suggestion
    const vm = this.getVmInstance(code, srcOrCode?.src);
```

File: src/lib/vm/vm.js [hunk 2]
```diff
@@ -2069,16 +2070,24 @@ export default class VM {
...
+        version, // may be undefined, meaning `latest`
+        undefined
+      );
+      if (!code) {
+        return code;
+      }
+    } else if (srcOrCode?.code) {
+      code = srcOrCode.code;
     }
+    console.log(code);
```
Comment (evgenykuzyakov) [inline src/lib/vm/vm.js:2090]: please remove
```suggestion
```

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]
Rebase to `dev` branch. Don't need to bump `package.json` for features

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,7 @@
   - For errors, expose `message`, `name` and `type`.
 - Fix `vm.depth` not being initialized.
 - Introduce `useMemo` hook. Similar to the React hook, it calculates a value and memoizes it, only recalculating when one of its dependencies changes.
+- Add support for VM.require when using redirectMap.
```
Comment (evgenykuzyakov) [inline CHANGELOG.md:14]: There is example below. Just move to the top of the list

**Comment by elliotBraem** [comment]
Made the requested changes, rebased to dev branch, but have now noticed something:

I get the error:

Screenshot 2023-10-07 at 3 11 25 PM

This isn't broken by my change, the same is visible when using a component w/ VM.require on the dev near social gateway:
https://dev.near.social/discover.near/widget/ProjectFeed


**Found the solution**

Need to bind this to VM during initGlobalFunctions

```javascript
      VM: {
        require: this.vmRequire.bind(this),
      },
```

Have added this to PR. It's all working as expected 👍

**Review by evgenykuzyakov — CHANGES_REQUESTED** [review]
I think we need to handle `getVmInstance` differently if the `src` not available. E.g. if the code is passed directly, then the src will be `null` or `undefined`. In this case we should store by the `code` instead.

Maybe we should store all VM instances in `this.vmInstances` by the code.

**Review by evgenykuzyakov — APPROVED** [review]

**Pull request merged**